### PR TITLE
Add MockTools for testing HTTP1ConnectionPool.StateMachine

### PR DESCRIPTION
**Pull request by fabianfett**

### CODE REVIEW

**Review by Lukasa — CHANGES_REQUESTED** [review]

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+            case .starting:
+                throw Errors.connectionNotIdle
+            case .http1(let leased, _):
+                if leased {
+                    throw Errors.connectionNotIdle
+                }
+            case .http2(_, let used):
+                if used > 0 {
+                    throw Errors.connectionNotIdle
+                }
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:195]: Do we need to transition to `closed` here on the happy path?

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+        self.connections.values
+            .filter { $0.isParked }
+            .min(by: { $0.lastIdle! < $1.lastIdle! })
+            .flatMap { .__testOnly_connection(id: $0.id, eventLoop: $0.eventLoop) }
+    }
+
+    func newestParkedConnection(for eventLoop: EventLoop) -> Connection? {
+        self.connections.values
+            .filter { $0.eventLoop === eventLoop && $0.isParked }
+            .sorted(by: { $0.lastIdle! > $1.lastIdle! })
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:244]: This is redundant with the `max(by:)` below
```suggestion
```

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+
+    func connections(on eventLoop: EventLoop) -> Int {
+        self.connections.values.filter { $0.eventLoop === eventLoop }.count
+    }
+
+    // MARK: Connection creation
+
+    mutating func createConnection(_ connectionID: Connection.ID, on eventLoop: EventLoop) throws {
+        guard self.connections[connectionID] == nil else {
+            throw Errors.connectionIDAlreadyUsed
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:257]: do we really need to distinguish this from `Errors.connectionExists`?

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+    mutating func finishExecution(_ connectionID: Connection.ID) throws {
+        guard var connection = self.connections[connectionID] else {
+            throw Errors.connectionNotFound
+        }
+
+        try connection.finishRequest()
+        self.connections[connectionID] = connection
+    }
+
+    mutating func randomStartingConnection() -> Connection.ID? {
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:365]: These are all for the fuzzing, presumably? Might be worth sticking them in an extension.

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+        case connectionIsNotExecuting
+        case connectionDoesNotFulfillEventLoopRequirement
+        case connectionBackoffTimerExists
+        case connectionBackoffTimerNotFound
+    }
+
+    private struct MockConnectionState {
+        typealias ID = HTTPConnectionPool.Connection.ID
+
+        enum State {
```
[the remark this answers is not in the document]
Comment (fabianfett) [reply Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:44]: We could make this `Hashable`, but TBH really his should be private. So the benefit of it being `Hashable` is not very large. :)

[the remark this answers is not in the document]
Comment (glbrntt) [reply Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift]: Agreed, it isn't large. I was thinking along the lines of killing some of the `switch`es in the various `is<State>` computed properties.

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 6]
```diff
@@ -0,0 +1,535 @@
...
+                if maxStreams < 1 {
+                    throw Errors.connectionDoesNotHaveHTTP2StreamAvailable
+                }
+                self.state = .http2(.inUse(maxConcurrentStreams: maxStreams, used: 1))
+
+            case .http2(.inUse(let maxStreams, let used)):
+                if let required = request.requiredEventLoop, required !== self.eventLoop {
+                    throw Errors.connectionDoesNotFulfillEventLoopRequirement
+                }
+                if used + 1 > maxStreams {
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:204]: Seems odd to do the addition twice.
```suggestion
                if used >= maxStreams {
```

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 7]
```diff
@@ -0,0 +1,535 @@
...
+    private var connections = [Connection.ID: MockConnectionState]()
+    private var backoff = Set<Connection.ID>()
+
+    init() {}
+
+    var parked: Int {
+        self.connections.values.filter { $0.isParked }.count
+    }
+
+    var leased: Int {
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:259]: Should this be `used` now?

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+            throw Errors.connectionNotFound
+        }
+
+        try connection.finishRequest()
+        self.connections[connectionID] = connection
+    }
+}
+
+extension MockConnectionPool {
+    mutating func randomStartingConnection() -> Connection.ID? {
```
Comment (glbrntt) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:414]: Why are the `random<State>Connection` functions `mutating`?

**Review by glbrntt — APPROVED** [review]
Looks good aside from a couple of trivial things

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+                if let required = request.requiredEventLoop, required !== self.eventLoop {
+                    throw Errors.connectionDoesNotFulfillEventLoopRequirement
+                }
+                self.state = .http1(.inUse)
+
+            case .http2(.idle(let maxStreams, false, _)):
+                if let required = request.requiredEventLoop, required !== self.eventLoop {
+                    throw Errors.connectionDoesNotFulfillEventLoopRequirement
+                }
+                if maxStreams < 1 {
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:195]: I don't think this state is real: a HTTP/2 connection is not allowed to have max streams less than 1.

File: Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift [hunk 10]
```diff
@@ -0,0 +1,535 @@
...
+                throw Errors.connectionIsNotExecuting
+
+            case .http1(.inUse):
+                self.state = .http1(.idle(parked: false, idleSince: .now()))
+
+            case .http2(.inUse(let maxStreams, let used)):
+                if used == 1 {
+                    self.state = .http2(.idle(maxConcurrentStreams: maxStreams, parked: false, lastIdle: .now()))
+                } else {
+                    self.state = .http2(.inUse(maxConcurrentStreams: maxStreams, used: used))
```
Comment (Lukasa) [inline Tests/AsyncHTTPClientTests/Mocks/MockConnectionPool.swift:226]: Shouldn't `used` be decremented here?

**Review by Lukasa — CHANGES_REQUESTED** [review]

**Review by Lukasa — APPROVED** [review]

**Pull request merged**